### PR TITLE
MODSOURCE-50 - Create endpoint to save a list of Records

### PR DESCRIPTION
**Pull request by RuslanLavrov**

### CODE REVIEW

File: ramls/source-record-storage.raml [hunk 1]
```diff
@@ -176,3 +176,29 @@ resourceTypes:
           body:
             text/plain:
               example: "Internal server error"
+  /recordsBulk:
```
Comment (KaterynaSenchenko) [inline ramls/source-record-storage.raml]: let's think of a different name for this endpoint

File: ramls/source-record-storage.raml [hunk 2]
```diff
@@ -176,3 +176,29 @@ resourceTypes:
...
+    post:
+      description: "sdf"
```
Comment (KaterynaSenchenko) [inline ramls/source-record-storage.raml]: add comprehensible description

File: ramls/source-record-storage.raml [hunk 3]
```diff
@@ -176,3 +176,29 @@ resourceTypes:
...
+      is: [validate]
+      body:
+        application/json:
+          type: recordCollection
+      responses:
+        204:
```
Comment (KaterynaSenchenko) [inline ramls/source-record-storage.raml]: why not 201? with body

File: mod-source-record-storage-server/src/main/java/org/folio/rest/impl/SourceStorageImpl.java [hunk 4]
```diff
@@ -304,6 +305,22 @@ public void postSourceStoragePopulateTestMarcRecords(TestMarcRecordsCollection e
     });
   }
 
+  @Override
+  public void postSourceStorageRecordsBulk(RecordCollection entity, Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    vertxContext.runOnContext(v -> {
+      try {
+        recordService.saveRecords(entity, tenantId)
+          .map((Response) PostSourceStorageRecordsResponse
```
Comment (KaterynaSenchenko) [inline mod-source-record-storage-server/src/main/java/org/folio/rest/impl/SourceStorageImpl.java]: use response for this endpoint

File: mod-source-record-storage-server/src/main/java/org/folio/rest/impl/SourceStorageImpl.java [hunk 5]
```diff
@@ -304,6 +305,22 @@ public void postSourceStoragePopulateTestMarcRecords(TestMarcRecordsCollection e
...
+            .respond201WithApplicationJson(entity, PostSourceStorageRecordsResponse.headersFor201()))
+          .otherwise(ExceptionHelper::mapExceptionToResponse)
+          .setHandler(asyncResultHandler);
+      } catch (Exception e) {
+        LOG.error("Failed to create a record", e);
```
Comment (KaterynaSenchenko) [inline mod-source-record-storage-server/src/main/java/org/folio/rest/impl/SourceStorageImpl.java]: update log message, you tried to create multiple records

File: mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java [hunk 6]
```diff
@@ -63,6 +65,25 @@
...
+  @Override
+  public Future<Boolean> saveRecords(RecordCollection recordCollection, String tenantId) {
+    Future<Boolean> resultFuture = Future.future();
+    ArrayList<Future> saveFutures = new ArrayList<>();
+
+    for (Record record : recordCollection.getRecords()) {
+      saveFutures.add(saveRecord(record, tenantId));
+    }
+    CompositeFuture compositeFuture = CompositeFuture.all(saveFutures);
+    compositeFuture.setHandler(ar -> {
```
Comment (KaterynaSenchenko) [inline mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java]: can we just map composite future? this kind of handling seems redundant

File: ramls/source-record-storage.raml (hunk 1 above)
Comment (Igor-Gorchakov) [inline ramls/source-record-storage.raml]: /recordsCollection

File: ramls/source-record-storage.raml [hunk 7]
```diff
@@ -176,3 +176,29 @@ resourceTypes:
...
+          body:
+            text/plain: !!null
```
Comment (Igor-Gorchakov) [inline ramls/source-record-storage.raml]: Please return records collection

File: mod-source-record-storage-server/src/test/java/org/folio/rest/impl/RecordApiTest.java [hunk 8]
```diff
@@ -973,4 +975,90 @@ public void shouldReturnSourceRecordWithAdditionalInfoOnGetBySpecifiedSnapshotId
...
+      .statusCode(HttpStatus.SC_CREATED)
+      .body("records*.snapshotId", everyItem(is(snapshot_1.getJobExecutionId())))
+      .body("records*.recordType", everyItem(is(record_1.getRecordType().name())))
+      .body("records*.rawRecord.content", notNullValue())
+      .body("records*.additionalInfo.suppressDiscovery", everyItem(is(false)));
+    async.complete();
+  }
+
+  @Test
+  public void shouldReturnBadRequestOnPostWhenNoRecordsPassedInBulk() {
```
Comment (Igor-Gorchakov) [inline mod-source-record-storage-server/src/test/java/org/folio/rest/impl/RecordApiTest.java]: Please remove "bulk" words

File: mod-source-record-storage-server/src/test/java/org/folio/rest/impl/RecordApiTest.java [hunk 9]
```diff
@@ -40,6 +41,7 @@
 
   static final String SOURCE_STORAGE_SOURCE_RECORDS_PATH = "/source-storage/sourceRecords";
   private static final String SOURCE_STORAGE_RECORDS_PATH = "/source-storage/records";
+  private static final String SOURCE_STORAGE_RECORDS_BULK_PATH = "/source-storage/recordsCollection";
```
Comment (Igor-Gorchakov) [inline mod-source-record-storage-server/src/test/java/org/folio/rest/impl/RecordApiTest.java]: SOURCE_RECORDS_COLLECTION_PATH

File: mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java [hunk 10]
```diff
@@ -63,6 +65,15 @@
       .compose(generation -> recordDao.saveRecord(record.withGeneration(generation), tenantId));
   }
 
+  @Override
+  public Future<Boolean> saveRecords(RecordCollection recordCollection, String tenantId) {
+    ArrayList<Future> saveFutures = new ArrayList<>();
+    recordCollection.getRecords()
+      .forEach(record -> saveFutures.add(saveRecord(record, tenantId)));
+
```
Comment (Igor-Gorchakov) [inline mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java]: remove blank line

**Comment by Igor-Gorchakov** [comment]
Update news.md

**Comment by Igor-Gorchakov** [comment]
Update module-descriptor.json

**Comment by Igor-Gorchakov** [comment]
Update module-descriptor.json

**Pull request merged**